src/openms/source/ANALYSIS/MAPMATCHING/FeatureGroupingAlgorithmKD.cpp

```diff
@@ -393,18 +397,95 @@ namespace OpenMS
 
   ClusterProxyKD FeatureGroupingAlgorithmKD::computeBestClusterForCenter_(Size i, vector<Size>& cf_indices, const vector<Int>& assigned, const KDTreeFeatureMaps& kd_data) const
   {
+    //Parameters how to use charge/adduct information
+    String merge_charge(param_.getValue("link:charge_merging").toString());
+    String merge_adduct(param_.getValue("link:adduct_merging").toString());
+
     // compute i's neighborhood, together with a look-up table
     // map index -> corresponding points
     map<Size, vector<Size> > points_for_map_index;
     vector<Size> neighbors;
     kd_data.getNeighborhood(i, neighbors, rt_tol_secs_, mz_tol_, mz_ppm_, true);
     Int charge_i = kd_data.charge(i);
+    const BaseFeature* f_i = kd_data.feature(i);
     for (vector<Size>::const_iterator it = neighbors.begin(); it != neighbors.end(); ++it)
     {
-      if (!assigned[*it] && kd_data.charge(*it) == charge_i)
+      //If the feature was already assigned, don't consider it at all!
+      if (assigned[*it])
+      {
+        continue;
+      }
+
+      if (merge_charge == "None")
+      {
+        if (kd_data.charge(*it) != charge_i)
+        {
+          continue;
+        }
+      }
+      //what to consider for linking with existing features _that have charge_. This ensures that we won't collect different non-zero charges.
+      else if (merge_charge == "Zero")
+      {
+        if (kd_data.charge(*it) != charge_i && kd_data.charge(*it) != 0)
+        {
+          continue;
+        }
+      }
+      else if (merge_charge == "All")
+      {
+        //we allow to merge all, keep case for parameter check
+      }
+      else
+      {
+        throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "WARNING! Invalid charge merging parameter! " + String(param_.getValue("link:charge_merging"))  + "\n");
+      }
+
+      //analogous adduct block
+      if (merge_adduct == "None")
       {
-        points_for_map_index[kd_data.mapIndex(*it)].push_back(*it);
+        //subcase 1: one has adduct, other not
+        if (kd_data.feature(*it)->metaValueExists("dc_charge_adducts") != f_i->metaValueExists("dc_charge_adducts"))
+        {
+          continue;
+        }
+        //subcase 2: both have adduct, but is it the same?
+        if (kd_data.feature(*it)->metaValueExists("dc_charge_adducts"))
+        {
+          if (EmpiricalFormula(kd_data.feature(*it)->getMetaValue("dc_charge_adducts")) != EmpiricalFormula(f_i->getMetaValue("dc_charge_adducts")))
```

Comment:
Do you need to create an EF for this? String comparison is more fragile, but might be sufficient here?

Reply:
Theoretically different gains/losses of charged/neutral adducts might result in the same end EF contained atoms. I don't know enough about the EF behavior, is the atom order designed to be deterministic?